src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
+    {
+        /// <summary>
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(ProjectId projectId, string projectFullPath, string fileFullPath, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// SourceFileInfo is random name I choose. we might want to chagne this to something like temporarySourceInfo,
+    /// InterimSourceInfo, GeneratedSourceInfo and etc.
+    /// 
+    /// none of info returned from property can be dynamic. it should always return same data
+    /// </summary>
+    internal abstract class SourceFileInfo
+    {
+        /// <summary>
+        /// file path of this source file. since this file can be virtual (doesn't actually exist in the file system),
+        /// file path doesn't need to point to actual physical file in the file system.
+        /// 
+        /// this can be used with something like editorconfig to find out right options for this source file.
+        /// </summary>
+        public abstract string FilePath { get; }
+
+        /// <summary>
+        /// return source code kind of this source file info
+        /// </summary>
+        public abstract SourceCodeKind SourceCodeKind { get; }
+
+        /// <summary>
+        /// provide current <see cref="TextLoader"/> and <see cref="IDocumentServiceProvider"/>
+        /// </summary>
+        /// <remarks>
+        /// we probably want to change this to Async and give in cancellation token. 
+        /// for now, it is not, since FileChange event is synchronous and doesn't provide cancellationToken
+        /// </remarks>
+        public abstract (TextLoader textLoader, IDocumentServiceProvider documentServiceProvider) GetTextLoaderAndDocumentServiceProvider();
+
+        /// <summary>
+        /// indicates that content of this source file has updated. new content will be retrieved by
+        /// TextLoader.LoadTextAndVersionAsync
+        /// 
+        /// I am going to start with simple one which just indicates changes. 
+        /// and not give in what old and new text/version were when event happened but just FilePath.
+        /// 
+        /// the event arg here (FilePath) is the original non source file full path given to
+        /// GetSourceFileInfoAsync
```

Comment:
This event also has to be abstract. If my code doesn't define it, I can't raise it.

Reply:
added OnUpdate method so that you can raise it.

Reply:
That works too 👍